src/sink/prometheus.rs

```diff
@@ -1,11 +1,30 @@
 //! Prometheus is a pull-based aggregation server
+//!
+//! This sink emits all cernan aggregations into prometheus according to the
+//! following method:
+//!
+//!   - SET -> gauge
+//!   - SUM -> counter
+//!   - QUANTILES -> summary 
+//!   - HISTOGRAM -> histogram
+//!
+//! All points are retained in a sliding window defined by
+//! `retain_limit`. Points will persist indefinately unless new points come
+//! behind them to push them out of the `retain_limit` window.
```

Comment:
Does `retain_limit` aggregation apply to `gauge` metrics as well?  If so, how does that work?

Reply:
It applies to all cernan aggregation methods, yes. If a prometheus aggregation should be treated differently do let me know. Otherwise, points persist indefinitely until a new point comes along to knock it out of place.

Reply:
As long as cernan is giving prometheus the most recent value of the gauge and not doing something weird like summing or averaging the last 10 points, it sounds harmless :-)

Reply:
What happens is we sum all of the values that we're storing: https://github.com/postmates/cernan/pull/306/files#diff-497dac92ece3082c6719be9ac7ca33e2R138

AddAssign for Telemetry is defined here: https://github.com/postmates/cernan/blob/da80ac412cc2bfccbea6b157c67f07f3744089a7/src/metric/telemetry.rs#L213

Add for Value is defined here: https://github.com/postmates/cernan/blob/da80ac412cc2bfccbea6b157c67f07f3744089a7/src/metric/telemetry.rs#L79

Reply:
The specific line you want is here: https://github.com/postmates/cernan/blob/da80ac412cc2bfccbea6b157c67f07f3744089a7/src/metric/telemetry.rs#L95

LHS (SET | SUM) + RHS (SET) = RHS (SET)

Reply:
Oh, I think I see what was confusing me.  This is specifically how to handle _aggregating_ multiple sources of a single gauge metric, right?  If so, the ideal case for prometheus would be not to aggregate them at all, but instead to export each value as a separate metric, differentiated by a label (probably `source`).  Then Prometheus can aggregate them according to its configured rules.  Hopefully I'm not totally lost here...

Reply:
I think we need to make a careful distinction between prometheus' aggregation and cernan Telemetry and how summation over Telemetry works. Points:

1. Prometheus aggregations do not exist until we hit `write_text` or `write_binary` at which point we convert from cernan's Telemetry into the appropriate serialized version. 
2. Cernan aggregations are SET, SUM, HISTOGRAM and SUMMARIZE, defined here: https://github.com/postmates/cernan/wiki/Data-Model#aggregation (HISTOGRAM is left off the documented list until https://github.com/postmates/cernan/issues/290 lands)

The comment above indicates how cernan aggregation methods sum up. In the prometheus sink we keep up to `retain_limit` time bins worth of Telemetry and then sum over them to get the Telemetry we'll report out to prometheus. A SUM will be SUM(x_0 + x_1 + ... x_retain_limit) for instance. A SET will be SET(x_retain_limit). 

All that said, I'm not sure what you mean to suggest by 

> If so, the ideal case for prometheus would be not to aggregate them at all, but instead to export each value as a separate metric, differentiated by a label (probably `source`).

What would the value of `source` be? Some configurations of cernan have a `source` defined in their tag map but not all so I don't imagine that's what you mean.

Reply:
> What would the value of source be? Some configurations of cernan have a source defined in their tag map but not all so I don't imagine that's what you mean.

`source` would be the name of the instance that sent that value to cernan, presumably.  Hostname of the source IP, unless there is a better name to use somewhere.

Reply:
> `source` would be the name of the instance that sent that value to cernan, presumably. Hostname of the source IP, unless there is a better name to use somewhere.

Got it. These things are only available by configuration, per https://github.com/postmates/cernan/wiki/Configuration#global-tags. Most users do set tags, but not all. 

This doesn't address what cernan should do in the case of a single cernan instance receiving metrics from a single source. Currently, Telemetry is kept according to timestamp -- update to `retain_limit` values -- and then aggregated as discussed.

Reply:
Actually I might be on the wrong track with the labeling part, at least as a default.  The prometheus statsd exporter doesn't do that, for instance.

However, the part where counters are a sum of only the most recent N values seems extremely wrong to me.  Here's a side by side comparison of sending some counter data to both cernan and an instance of statsd_exporter:

```sh
#!/usr/bin/env bash
# Send some counter values to cernan and statsd_exporter and compare the result

metric_name="test.$$.bar"
values=(5 5 5 5 5 5 5 5 5 5 2 2 2 2 2 5 5 2 5 2 5 5 5 2 2 2 2 1)

echo "Sending ${#values[@]} values:"
echo -n "$metric_name "

for n in "${values[@]}"; do
  echo "$metric_name:$n|c"|nc -w1 -u 127.0.0.1 8125
  echo "$metric_name:$n|c"|nc -w1 -u 127.0.0.1 9125
  echo -n "$n "
done
echo

echo "Cernan result: "
curl -s http://127.0.0.1:8080/metrics|grep "$metric_name"
echo

echo "statsd_exporter result: "
curl -s http://127.0.0.1:9102/metrics|grep "$metric_name"
echo
```

After running the above:
```
Sending 28 values:
test.21981.bar 5 5 5 5 5 5 5 5 5 5 2 2 2 2 2 5 5 2 5 2 5 5 5 2 2 2 2 1 
Cernan result: 
# TYPE test_21981_bar counter
test_21981_bar 31

statsd_exporter result: 
# HELP test_21981_bar_counter Metric autogenerated by statsd_exporter.
# TYPE test_21981_bar_counter counter
test_21981_bar_counter 103
```

Assuming that the metric in question represents something like a total http requests counter, and ignoring the differing metric name (I don't know why stastd_exporter adds _counter when the data type is already explicitly a counter), how does cernan's final value of 31 make any sense?

Reply:
> how does cernan's final value of 31 make any sense?

That's what this review is for. :) Once upon a time there was an ask to do sliding window – https://github.com/postmates/cernan/issues/285 – but that conversation dead-ended. Now we can get a fix on the semantics. 

Okay, so it looks like, so far:

 - SET should always emit the last value in, forever 
 - SUM should always emit the running sum of values in, forever 

How about HISTOGRAM and SUMMARIZE?

Reply:
I'm writing up some test cases for these to figure out what to do, sorry for dropping the ball on you again.

Reply:
@benley cool, no worries. Let me know if I can help.

Reply:
When combined with the new histogram feature in #320 the histogram and summary features we'll be looking for seem to be in good shape.

Reply:
@benley Okay, to confirm:

* SET should always emit the last value in, forever 
* SUM should always emit the running sum of values in, forever 
* SUMMARIZE should emit a moving window quantiles summary, `retain_limit` total seconds across
* HISTOGRAM should emit a moving window histogram summary, `retain_limit` seconds across

Reply:
hmm, I'm still not so sure about the retain_limit part.  In prometheus land, the counters used in histograms and summaries would get reset to 0 when the process exporting metrics restarts.  Cernan doesn't restart when the metric sources restart, but we're also not relying on those counter resets to spot downtime or app restarts.  I think it will make more sense if HISTOGRAMs emit values since forever, like SET and SUM, as the data remains coherent that way. [The prometheus docs](https://prometheus.io/docs/concepts/metric_types/#histogram) seem to concur.

Summaries are trickier, since quantile estimation necessarily involves sliding time windows.  There isn't going a universally applicable decay time, so unless there is a way to configure the retention limit, an arbitrary value will have to do.  Is there any way to make it be a window of N minutes rather than the N most recent values?

Reply:
> Cernan doesn't restart when the metric sources restart

That'd be a neat trick, but for something like statsd source we'd have to fake it with a liveness timer. 

> I think it will make more sense if HISTOGRAMs emit values since forever, like SET and SUM, as the data remains coherent that way. The prometheus docs seem to concur.

Easily done. 

> There isn't going a universally applicable decay time, so unless there is a way to configure the retention limit, 

What would you be looking for? Configuration per time series? 

> an arbitrary value will have to do. Is there any way to make it be a window of N minutes rather than the N most recent values?

Sure. Every flush pulse we can inspect the summaries and slide the underlying window or windows, if we want to make them configurable per series in some fashion.

Reply:
Configuration per time series would be cool but I'd settle for a time-based window for the quantiles instead of a count-based one.

One thing I haven't checked but should have: are the `<summaryname>_sum`  and `<summaryname>_count` metrics totals since forever?  If not, can you make them that away?

Reply:
> Configuration per time series would be cool but I'd settle for a time-based window for the quantiles instead of a count-based one.

I can make that happen, sure. 

> One thing I haven't checked but should have: are the `<summaryname>_sum` and `<summaryname>_count` metrics totals since forever? If not, can you make them that away?

Excellent question. So for SUMMARIZE we're setting the sample sum using value.sample_sum (for instance:  https://github.com/postmates/cernan/blob/b2ef685e5e28c83613a68a4e2f4d699376df0480/src/sink/prometheus.rs#L344) and that's defined [here](https://github.com/postmates/cernan/blob/b2ef685e5e28c83613a68a4e2f4d699376df0480/src/metric/telemetry.rs#L696), so we're bouncing out to CKMS' sum: https://github.com/postmates/quantiles/blob/1f4232b7b968d0688c1d71f5bdfcb12c7f49a69f/src/ckms.rs#L189 The nice thing here is that sum is totally additive, even during AddAssign: https://github.com/postmates/quantiles/blob/1f4232b7b968d0688c1d71f5bdfcb12c7f49a69f/src/ckms.rs#L70. 

Hmm, now sample count is https://github.com/postmates/cernan/blob/b2ef685e5e28c83613a68a4e2f4d699376df0480/src/metric/telemetry.rs#L174 which for SUMMARIZE bounces to https://github.com/postmates/quantiles/blob/1f4232b7b968d0688c1d71f5bdfcb12c7f49a69f/src/ckms.rs#L337 which is also only added to, so we're golden. 

What'll be a little weird, since we're going to make the samples happen over a time window is we'll need to keep track of previous windows' sum and count. Which, easily done, just need to watch out for that in review.

Reply:
@benley 671841013968cad301430d8bf257a6b6eed9c507 + fbc9baaba30b2ff6e53dc1ec3372ece1276e09fe + bba51840e32f3e96e1d379c3318712733e0fd6ee ought to be what we're looking for.